### PR TITLE
feat: mining control for testing providers

### PR DESCRIPTION
**Pull request by johnson2427**
### What I did
Adding mining capability

fixes: #344

### How I did it
Created abstract method `mine` to the `TestProviderAPI` class in `src/ape/api/providers.py`
Added the `mine`, `enable_automining`, and `disable_automining` method to the `ChainManager`  in `src/ape/managers/chain.py`.
Added `mine` method into the `LocalNetwork` class in `src/ape_test/providers.py`

### Checklist

- [ ] Passes all linting checks (pre-commit and CI jobs)
- [ ] New test cases have been added and are passing
- [ ] Documentation has been updated
- [ ] PR title follows [Conventional Commit](https://www.conventionalcommits.org/en/v1.0.0/) standard (will be automatically included in the changelog)

### CODE REVIEW

File: src/ape/api/providers.py [hunk 1]
```diff
@@ -600,6 +600,15 @@ def set_timestamp(self, new_timestamp: int):
             int: The new timestamp.
         """
 
+    @abstractmethod
+    def mine(self, num_blocks: int):
```
Comment (unparalleled-js) [inline src/ape/api/providers.py:604]: Should this default to `1`, you think?

File: src/ape/api/providers.py [hunk 2]
```diff
@@ -600,6 +600,15 @@ def set_timestamp(self, new_timestamp: int):
...
+        """
+        Args:
+            num_blocks (int): The number of blocks allotted to mine
```
Comment (unparalleled-js) [inline src/ape/api/providers.py:607]: I think we need a newline between the args and returns blocks or else the docs build funny

File: src/ape/managers/chain.py [hunk 3]
```diff
@@ -438,3 +438,15 @@ def _get_test_provider(self, method_name: str) -> TestProviderAPI:
             )
 
         return provider
+
+    def mine(self, num_blocks: int = 1, timestamp: Optional[int] = None) -> str:
+        if timestamp:
+            self.pending_timestamp = timestamp
+        self.provider.mine(num_blocks)
+        return f'Mining: {num_blocks}, {timestamp}'
```
Comment (unparalleled-js) [inline src/ape/managers/chain.py:446]: I am a little confused by this return type. Should we just return the new head instead? (`BlockAPI`)

Comment (unparalleled-js) [reply src/ape/managers/chain.py:446]: Or maybe we don't need a return type here?

**Comment by unparalleled-js** [comment]
In `tests/integration/test_chain/py`, there are several spots where we are making a `transfer()` just for the sake of mining (because of automine).  Can we instead replace those spots with a call `mine()`? That way, we are also testing `mine()` itself.

**Comment by unparalleled-js** [comment]
Also a test for disabling automine would be a nice!

File: src/ape/api/providers.py (hunk 1 above)
Comment (fubuloubu) [reply src/ape/api/providers.py:604]:
```suggestion
    def mine(self, num_blocks: int = 1):
```

File: src/ape/managers/chain.py (hunk 3 above)
Comment (fubuloubu) [reply src/ape/managers/chain.py:446]: Yeah agreed, no need to have a return type here, or print out anything for display.

Maybe debug log might be useful... :thinking: @unparalleled-js thoughts?

File: src/ape/managers/chain.py [hunk 4]
```diff
@@ -438,3 +438,15 @@ def _get_test_provider(self, method_name: str) -> TestProviderAPI:
...
+
+    def enable_automining(self):
+        pass
+
+    def disable_automining(self):
+        pass
```
Comment (fubuloubu) [inline src/ape/managers/chain.py:452]: no need to include these yet

File: src/ape_test/providers.py [hunk 5]
```diff
@@ -76,6 +76,11 @@ def revert(self, snapshot_id: SnapshotID):
     def set_timestamp(self, new_timestamp: int):
         self._tester.time_travel(new_timestamp)
 
+    def mine(self, num_blocks: int):
```
Comment (fubuloubu) [inline src/ape_test/providers.py:79]:
```suggestion
    def mine(self, num_blocks: int = 1):
```

File: src/ape_test/providers.py [hunk 6]
```diff
@@ -76,6 +76,11 @@ def revert(self, snapshot_id: SnapshotID):
...
+        for i in range(num_blocks):
+            txn = TransactionAPI()
+            self.send_transaction(txn)
```
Comment (fubuloubu) [inline src/ape_test/providers.py:82]: So `eth-tester` has this functionality already:
https://github.com/ethereum/eth-tester/blob/master/eth_tester/main.py#L348

```suggestion
        self._tester.mine_blocks(num_blocks)
```

**Review by fubuloubu — CHANGES_REQUESTED** [review]
A couple of comments, but overall looking pretty good!

Address these and we can probably merge, and then also target `ape-hardhat` for an update

File: src/ape/managers/chain.py (hunk 3 above)
Comment (unparalleled-js) [reply src/ape/managers/chain.py:446]: A debug log could be useful but hard to say. Does not hurt though, it is already a noisy stream atm.

File: src/ape/api/providers.py [hunk 7]
```diff
@@ -600,6 +600,15 @@ def set_timestamp(self, new_timestamp: int):
...
+        Returns:
```
Comment (fubuloubu) [inline src/ape/api/providers.py:608]:
```suggestion

        Returns:
```

File: src/ape/managers/chain.py [hunk 8]
```diff
@@ -438,3 +438,15 @@ def _get_test_provider(self, method_name: str) -> TestProviderAPI:
             )
 
         return provider
+
+    def mine(self, num_blocks: int = 1, timestamp: Optional[int] = None) -> str:
```
Comment (fubuloubu) [inline src/ape/managers/chain.py:442]:
```suggestion
    def mine(self, num_blocks: int = 1, timestamp: Optional[int] = None):
```

File: src/ape/managers/chain.py (hunk 3 above)
Comment (fubuloubu) [inline src/ape/managers/chain.py:446]:
```suggestion
```

File: src/ape/managers/chain.py (hunk 4 above)
Comment (fubuloubu) [reply src/ape/managers/chain.py:452]:
```suggestion
```

Comment (johnson2427) [reply src/ape/managers/chain.py:452]: deleted this

File: src/ape_test/providers.py (hunk 5 above)
Comment (johnson2427) [reply src/ape_test/providers.py:79]: added this, we had it in the chain.py, just forgot to add it to the providers

File: src/ape_test/providers.py (hunk 6 above)
Comment (johnson2427) [reply src/ape_test/providers.py:82]: correction made

File: src/ape/api/providers.py (hunk 7 above)
Comment (johnson2427) [reply src/ape/api/providers.py:608]: done

File: src/ape/managers/chain.py (hunk 8 above)
Comment (johnson2427) [reply src/ape/managers/chain.py:442]: done

File: src/ape/managers/chain.py (hunk 3 above)
Comment (johnson2427) [reply src/ape/managers/chain.py:446]: if me commenting on all of these is annoying, let me know. I just use this as a check list

File: src/ape/api/providers.py [hunk 9]
```diff
@@ -600,6 +600,16 @@ def set_timestamp(self, new_timestamp: int):
             int: The new timestamp.
         """
 
+    @abstractmethod
+    def mine(self, num_blocks: int = 1):
+        """
+        Args:
+            num_blocks (int): The number of blocks allotted to mine
```
Comment (unparalleled-js) [inline src/ape/api/providers.py:607]:
```suggestion
            num_blocks (int): The number of blocks allotted to mine. Defaults to ``1``.
```

File: src/ape/api/providers.py [hunk 10]
```diff
@@ -600,6 +600,16 @@ def set_timestamp(self, new_timestamp: int):
...
+
+        Returns:
```
Comment (unparalleled-js) [inline src/ape/api/providers.py:609]: You can delete the `Returns:` block now

File: src/ape/managers/chain.py [hunk 11]
```diff
@@ -438,3 +438,8 @@ def _get_test_provider(self, method_name: str) -> TestProviderAPI:
             )
 
         return provider
+
+    def mine(self, num_blocks: int = 1, timestamp: Optional[int] = None) -> None:
```
Comment (fubuloubu) [inline src/ape/managers/chain.py:442]: Add docstrings for this method

**Review by fubuloubu — APPROVED** [review]
To fix the commit message checks, just do `git rebase -I main` and edit all the commit message s to be lowercase `feat:`. Also the first word after `:` must be lowercase as well, for some reason.
image

The type checking error, you have to add the `mine` method to `ProviderAPI` as well, using the `raises_not_implemented` decorator like in: https://github.com/ApeWorX/ape/pull/406

I think you have to rebase it anyways for it to merge properly too

Otherwise, looks great! Just one docs thing and good to merge!

**Review by unparalleled-js — APPROVED** [review]
LGTM!

**Review by unparalleled-js — APPROVED** [review]

**Pull request merged**